extensions/wikia/Rail/RailController.class.php

```diff
@@ -8,139 +8,159 @@ class RailController extends WikiaController {
-					$railModule[0], /* Controller */
-					$railModule[1], /* Method */
-					$railModule[2] /* array of params */
-				);
-			}
+		global $wgUseSiteJs, $wgAllowUserJs, $wgTitle, $wgAllInOne, $wgUser;
+		$title = Title::newFromText(
+			$this->request->getVal( 'articleTitle', null ),
+			$this->request->getInt( 'namespace', null )
+		);
+
+		if ( !( $title instanceof Title ) ) {
+			return;
+		}
 
-			$this->railLazyContent .= Html::element('div', ['id' => 'WikiaAdInContentPlaceHolder']);
+		// override original wgTitle from title given in parameters
+		// we cannot use wgTitle that is created on by API because it's broken on wikis without '/wiki' in URL
+		// https://wikia-inc.atlassian.net/browse/BAC-906
+		$oldWgTitle = $wgTitle;
+		$wgTitle = $title;
+		$assetManager = AssetsManager::getInstance();
+		$railModules = $this->filterModules(
+			( new BodyController )->getRailModuleList(),
+			self::FILTER_LAZY_MODULES
+		);
+		$this->railLazyContent = '';
+
+		krsort( $railModules );
+
+		// TODO XW-2760 remove after experiment is done
+		$isAdMixExperimentEnabled = $this->request->getVal( 'isAdMixExperimentEnabled', false );
+
+		if ( $isAdMixExperimentEnabled ) {
+			// ad slot NATIVE_TABOOLA_RAIL
+			unset( $railModules[1435] );
+			// ad slot LEFT_SKYSCRAPER_2
+			unset( $railModules[1100] );
+
+			// copied from RecirculationHooks::onGetRailModuleList
+			$recirculationModulePosition = $wgUser->isAnon() ? 1305 : 1285;
+			unset( $railModules[$recirculationModulePosition] );
+
+			array_push( $railModules, [ 'AdMixExperiment', 'recirculationAndAdPlaceholder' ] );
+		}
 
-			$this->css = $sassFiles = [];
-			foreach (array_keys($this->app->wg->Out->styles) as $style) {
-				if ($wgAllInOne && $assetManager->isSassUrl($style)) {
-					$sassFiles[] = $style;
-				} else {
-					$this->css[] = $style;
-				}
-			}
+		foreach ( $railModules as $railModule ) {
+			$this->railLazyContent .= $this->app->renderView(
+				$railModule[0], /* Controller */
+				$railModule[1], /* Method */
+				$railModule[2] /* array of params */
+			);
+		}
 
-			if (!empty($sassFiles)) {
-				$excludeScss = (array) $this->getRequest()->getVal('excludeScss', []);
-				$sassFilePath = (array) $assetManager->getSassFilePath($sassFiles);
-				$includeScss = array_diff($sassFilePath, $excludeScss);
+		// ad mix experiment uses a wrapper to group recirculation and ad placeholder
+		if ( !$isAdMixExperimentEnabled ) {
+			$this->railLazyContent .= Html::element( 'div', [ 'id' => 'WikiaAdInContentPlaceHolder' ] );
+		}
 
-				// SUS-771: Log any duplicate CSS that rail modules try to load but are already loaded by Oasis skin
-				$duplicateScss = array_intersect( $sassFilePath, $excludeScss );
-				if ( count( $duplicateScss ) ) {
-					Wikia\Logger\WikiaLogger::instance()->info( 'SUS-771', [
-						'styles' => json_encode( $duplicateScss )
-					] );
-				}
+		$this->css = $sassFiles = [];
+		foreach ( array_keys( $this->app->wg->Out->styles ) as $style ) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this code to rely on RequestContext instead of context-dependent globals. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)